### PR TITLE
ENFClient v2 - ExposureDetectionTracker without identifier (EXPOSUREAPP-3540)

### PR DESCRIPTION
**Pull request by BMItr**
This PR implements one additional part of the new ExposureWindowMode (#1584 ).

We don't have a token anymore and can't provide one to the ExposureDetectionTracker.
This PR implements the opportunity to use ExposureDetectionTracker with & without identifier.
Note: ExposureDetectionTracker was CalculationTracker some days ago (just to prevent confusion).

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ENFClient.kt [hunk 1]
```diff
@@ -40,19 +40,8 @@ class ENFClient @Inject constructor(
...
-
-        return if (keyFiles.isEmpty()) {
-            Timber.d("No key files submitted, returning early.")
-            true
-        } else {
-            Timber.d("Forwarding %d key files to our DiagnosisKeyProvider.", keyFiles.size)
-            exposureDetectionTracker.trackNewExposureDetection(token)
-            diagnosisKeyProvider.provideDiagnosisKeys(keyFiles, configuration, token)
-        }
+        // NO-OP
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ENFClient.kt:43]: Maybe `throw UnsupportedOperationException()` until this is removed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt [hunk 2]
```diff
@@ -93,34 +94,16 @@ class DefaultExposureDetectionTracker @Inject constructor(
         }
     }
 
-    override fun finishExposureDetection(identifier: String, result: Result) {
-        Timber.i("finishExposureDetection(token=%s, result=%s)", identifier, result)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt:97]: Please re-add a log statement directly at function start, the following methods using `updateSafely` are async, if there is any issue there we'd still like to know that the function was called, but failed somewhere later.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt [hunk 3]
```diff
@@ -134,6 +117,54 @@ class DefaultExposureDetectionTracker @Inject constructor(
         }
     }
 
+    private fun finishExposureDetectionWithoutIdentifier(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt:120]: Would an extension function improve readability here? 

Something like `fun Map<String, TrackedExposureDetection>.findOrCreateByIdentifier(identifier: String?)`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt [hunk 4]
```diff
@@ -134,6 +117,54 @@ class DefaultExposureDetectionTracker @Inject constructor(
...
+    ) {
+        Timber.d(
+            "finishExposureDetectionWithoutIdentifier(): " +
+                "Get identifier of newest unfinished detection or create new identifier"
+        )
+        val identifier = map
+            .map { it.value }
+            .filter { it.finishedAt == null }
+            .maxByOrNull { it.startedAt.millis }
+            ?.identifier ?: kotlin.run {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt:132]: A bit difficult to read. How could we make this easier on the :eyes:

Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt]: good point. 
_?.identifier ?: kotlin.run_ is out.
readability improved

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt (hunk 3 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt]: I changed it to:
private fun Map<String, TrackedExposureDetection>.findUnfinishedOrCreateIdentifier(): String {

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt (hunk 2 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/detectiontracker/DefaultExposureDetectionTracker.kt:97]: added.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ENFClient.kt (hunk 1 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ENFClient.kt]: throwing UnsupportedOperationException() now

**Review by kolyaopahle — APPROVED** [review]
LGTM 👍 all handleable changes seem solved, lets get this merged to get the main branch ready, tests can be fixed later before we merge into release

**Pull request merged**